ENH: Create a Calibrations class for eyetracking data

scott-huberty opened the pull request:
#### Reference issue
I forgot to open a ticket before creating a PR (apologies!):

This PR is based on my discussions with @larsoner and @britta-wstnr regarding my current GSoC project. For the first week, the goal was to extend functionality to reading and storing some meta-information from eye-tracking recordings, that will be needed for work down the line in this GSoC project.

#### What does this implement/fix?

This PR creates a `Calibration` class.

- [Here](https://output.circle-artifacts.com/output/job/88665d35-0ec7-4e6a-95fa-f6f714d5cd1d/artifacts/0/html/auto_tutorials/preprocessing/90_eyetracking_data.html) is the updated eye-tracking tutorial.

- [Here](https://output.circle-artifacts.com/output/job/88665d35-0ec7-4e6a-95fa-f6f714d5cd1d/artifacts/0/html/generated/mne.preprocessing.eyetracking.Calibration.html) is the documentation for the `Calibration` class.

- [Here](https://output.circle-artifacts.com/output/job/88665d35-0ec7-4e6a-95fa-f6f714d5cd1d/artifacts/0/html/generated/mne.preprocessing.eyetracking.read_eyelink_calibration.html#mne.preprocessing.eyetracking.read_eyelink_calibration) is the documentation for `read_eyelink_calibration`

- [Here](https://app.codecov.io/gh/mne-tools/mne-python/pull/11719?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=mne-tools) is the codecov report. 
##### Minimally Working Example:

```
fname = mne.datasets.testing.datapath() / "eyetrack" / "test_eyelink.asc"
calibrations = mne.preprocessing.eyetracking..read_eyelink_calibration(fname, 
                                               screen_size=(.531, .298),
                                               screen_distance=.065,
                                               screen_resolution=(1920, 1080))
```

`print(calibrations[0])  # show info for the left eye Calibration`

```
Calibration |
  onset: 0 seconds
  model: HV13
  eye: left
  average error: 0.3 degrees
  max error: 0.9 degrees
  screen size: (0.531, 0.298) meters
  screen distance: 0.065 meters
  screen resolution: (1920, 1080) pixels
```


#### Additional information

##### A couple of notes: 

- `screen_size`, `screen_distance`, and `screen_resolution` are not available in the ASCII files but can be optionally provided by the user.

- [This draft](https://bids-specification--1128.org.readthedocs.build/en/1128/modality-specific-files/eye-tracking.html) from the [WIP eye-tracking BIDS specificaton](https://github.com/bids-standard/bids-specification/pull/1128) was used to guide some of the decisions regarding the fields to include by default in the `Calibration` class. 


#### TODOS:

- [x] Fix failing tests
- [x] increase `codecov` of the changes introduce in this PR
- [x] move `mne.io.eyelink.calibration` to `mne.preprocessing.eyetrack.calibration`, since the `Calibrations` class is not specific to `eyelink`.
- [x] create a `plot` method for the `Calibration` class.
- [x] Add the a demonstration of reading and plotting the calibration to the current eye-tracking tutorial
- [x] Take a look at `Calibration['points']` to see if it can be split into separate `keys` such that it is easier to work with / extend to future I/O functions for other manufacturers.

larsoner commented on mne/io/eyelink/calibration.py:
```diff
@@ -0,0 +1,203 @@
+
+
+@fill_doc
+class Calibrations(list):
```
Rather than `Calibrations` I think it would be nicer to have `Calibration` that subclasses `dict` (basically just a `dict` with a nice `__repr__`) and have `read_calibrations` return `list of Calibration`

larsoner replied:
Oh I didn't read far enough that `Calibrations` does have `Calibration` as its entries... I'm not sure there's enough value added to this `Calibrations` class over just `Calibration`. Let's stick with bare `list of Calibration` as the return type to keep things simpler

scott-huberty replied:
Ok!

britta-wstnr commented:
Great, @scott-huberty ! Let me know once this is ready to be tested on additional data.

britta-wstnr commented on mne/io/eyelink/_utils.py:
```diff
@@ -0,0 +1,130 @@
+
+
+def _find_recording_start(lines):
+    """Return the first START line in an eyelink ASCII file.
```
Do we want to / have any obligations to use the correct product names here? Then it would be EyeLink.

britta-wstnr commented on mne/io/eyelink/calibration.py:
```diff
@@ -0,0 +1,203 @@
+        can be passed with the maximum error for each eye. For example,
+        {'left': 0.5, 'right': 0.6}.
+    points: ndarray
+        a 2D numpy array, which are the data for each calibration point.
+        Each row contains the x and y pixel-coordinates of the actual gaze position
```
```suggestion
        A 2D numpy array containing the data for each calibration point.
        Each row contains the x and y pixel-coordinates of the actual gaze position
```

britta-wstnr commented on mne/io/eyelink/calibration.py:
```diff
@@ -0,0 +1,203 @@
+        a horizontal only 3-point calibration, or HV3 for a horizontal and
+        vertical 3-point calibration.
+    eye: str
+        the eye that was calibrated. For example, 'left',
+        'right', or 'both'.
```
```suggestion
        The eye that was calibrated. For example, 'left',
        'right', or 'both'.
```

britta-wstnr reviewed (commented):
Some small stuff my _gaze_ tripped upon when looking at the PR - will do a more thorough review after tests pass 🙂

scott-huberty commented:
> Some small stuff my _gaze_ tripped upon when looking at the PR - will do a more thorough review after tests pass 🙂

Thanks! since I made quite a bit of changes locally, I just made your suggested changes locally on top of my changes, (instead of comitting them here and rebasing). See the commit with title `FIX: Incorporated Britta's suggested changes`

scott-huberty commented:
Maintenance note: After some refactoring and adding a `plot` method, I edited my initial message at the top to reflect this. I've included an example of the calibration plot, there.

mscheltienne commented:
The plot looks great, it could be nice to have the average and max error written in a corner.

scott-huberty commented:
> The plot looks great, it could be nice to have the average and max error written in a corner.

Good idea. I'll give it a shot!

larsoner commented:
I would nest the `from ...preprocessing.eyetracking.calibration import Calibration` in whatever function actually needs the `Calibration` class

larsoner commented on mne/io/eyelink/_utils.py:
```diff
@@ -0,0 +1,107 @@
+        A list containing one or more Calibration instances,
+        one for each calibration that was recorded in the eyelink ASCII file
+        data.
+    """
```
Like this, and remove it from the top
```suggestion
    """
    from ...preprocessing.eyetracking.calibration import Calibration
```

larsoner commented on mne/io/eyelink/_utils.py:
```diff
@@ -0,0 +1,107 @@
+
+import re
+
+from ...preprocessing.eyetracking.calibration import Calibration
+
```
```suggestion
```

scott-huberty commented:
> I would nest the `from ...preprocessing.eyetracking.calibration import Calibration` in whatever function actually needs the `Calibration` class

That did it, thanks!

scott-huberty commented:
It seems like inherited methods are being built by sphinx `autosumary`: [here is the Calibration doc page](https://output.circle-artifacts.com/output/job/1966e2e3-460d-436e-b36e-69a2f778e7c1/artifacts/0/html/generated/mne.preprocessing.eyetracking.Calibration.html)

And this is triggering some `sphinx` warnings during the build.

In `mne/doc/preprocessing.rst` we specify `:no-inherited-members:`, but this doesn't seem to solve the problem:


```
:py:mod:`mne.preprocessing.eyetracking`:

.. currentmodule:: mne.preprocessing.eyetracking

.. automodule:: mne.preprocessing.eyetracking
   :no-members:
   :no-inherited-members:

.. autosummary::
   :toctree: generated/

   Calibration
   set_channel_types_eyetrack
   ```

@larsoner / @britta-wstnr do you have any idea of what I'm doing wrong here?

larsoner commented:
... and also `"a shallow copy of od"`

The short explanation is that warnings in doc build are treated as errors. So if you search "warning" in the circle log you'll see these. And we can ignore them.

For the copy one it would actually be better to override the method using a deep copy like we do for Info and other objects that subclass `__dict__`. Basically in MNE there is an expectation that `.copy()` operates like a deep copy so we should probably continue that behavior here

scott-huberty commented:
> ... and also `"a shallow copy of od"`
> 
> The short explanation is that warnings in doc build are treated as errors. So if you search "warning" in the circle log you'll see these. And we can ignore them.
> 
> For the copy one it would actually be better to override the method using a deep copy like we do for Info and other objects that subclass `__dict__`. Basically in MNE there is an expectation that `.copy()` operates like a deep copy so we should probably continue that behavior here

Ok! Adding those phrases to `conf.py` and I will over ride the `copy` method.


What do you think we should do about the [warning triggered by the move_to_end inherited method](https://app.circleci.com/pipelines/github/mne-tools/mne-python/19215/workflows/a42c34fa-7d43-47f9-aca3-9bf2a3320aad/jobs/55056?invite=true#step-140-1167)?:

```
WARNING: [numpydoc] Validation warnings while processing docstring for 'mne.preprocessing.eyetracking.Calibration.move_to_end':
  GL02: Closing quotes should be placed in the line after the last text in the docstring (do not close the quotes in the same line as the text, or leave a blank line between the last text and the quotes)
  PR01: Parameters {'key', 'last'} not documented
```

scott-huberty commented:
> There should be a similar ignore somewhere for numpydoc. Take a look in conf.py and see if it's in there, if you can't find anything quickly I can look

I thought I found what I needed in `numpy_doc_show_inherited_class_members`, which I set to False:

```
numpydoc_show_inherited_class_members = {
    "mne.SourceSpaces": False,
    "mne.Forward": False,
    "mne.preprocessing.eyetracking.Calibration": False,
}
```

But [the warning is still raised,](https://app.circleci.com/pipelines/github/mne-tools/mne-python/19223/workflows/442fee67-dd99-472b-b258-f064d30b5230/jobs/55073?invite=true#step-140-1167) and the documentation[ is still clearly being built for inherited methods](https://output.circle-artifacts.com/output/job/1f4764cc-a566-42a8-8a90-04c60f30540c/artifacts/0/html/generated/mne.preprocessing.eyetracking.Calibration.html)


the search continues...

larsoner commented:
For no inherited members see how we document the Forward class in doc/, you need to change the autosummary template that gets used

larsoner commented:
https://github.com/mne-tools/mne-python/blob/main/doc/forward.rst?plain=1#L7

scott-huberty commented:
> For no inherited members see how we document the Forward class in doc/, you need to change the autosummary template that gets used

Thanks!

Almost there.. There's a [failure in the M1 tests](https://github.com/mne-tools/mne-python/pull/11719/checks?check_run_id=14055763038) and I am trying to figure out if it is specific to my PR:

```
mne.viz.backends.tests.test_abstract.test_widget_abstraction_pyvistaqt[pyvistaqt]

mne/viz/backends/tests/test_abstract.py:119: in test_widget_abstraction_pyvistaqt
    _do_widget_tests(backend)
mne/viz/backends/tests/test_abstract.py:28: in _do_widget_tests
    window = backend._AppWindow()
mne/viz/backends/_qt.py:707: in __init__
    _MNEMainWindow.__init__(self, size=size)
mne/viz/backends/_qt.py:698: in __init__
    MainWindow.__init__(self, parent=parent, title=title, size=size)
/opt/homebrew/lib/python3.10/site-packages/pyvistaqt/window.py:23: in __init__
    QMainWindow.__init__(self, parent=parent)
E   TypeError: _Widget.__init__() got an unexpected keyword argument 'parent'
```

larsoner commented:
No you can ignore that, I'll fix it tomorrow

scott-huberty commented:
Alright @larsoner @britta-wstnr  I think this is ready for review!

I'll open a ticket tomorrow for the next preprocessing function I'd like to work on. This work isn't a necessary pre-requisite for it so I won't be blocked, i.e no rush, feel free to review this when you have the time.


Here are the Built docs for:
- [read_eyelink_calibration](https://output.circle-artifacts.com/output/job/4db7fd21-9ea5-4edf-94bc-d5f3feceb705/artifacts/0/html/generated/mne.io.read_eyelink_calibration.html)
- [Calibration class](https://output.circle-artifacts.com/output/job/4db7fd21-9ea5-4edf-94bc-d5f3feceb705/artifacts/0/html/generated/mne.preprocessing.eyetracking.Calibration.html)

Here[ is the codecov report](https://app.codecov.io/gh/mne-tools/mne-python/pull/11719?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=mne-tools)

I updated my message at the top with figures generated from the updated `plot` method.

britta-wstnr commented:
Great @scott-huberty ! I'll give it a try no my EyeLink data as well and report back! 🙂

mscheltienne commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+
+    Parameters
+    ----------
+    filename : str
+        Path to the eyelink file (.asc).
```
This argument is usually called `fname` in the other IO functions.

scott-huberty replied:
> A couple of minor comments after gazing through the documentation, which can be applied to other docstrings too. Looks great!

Thanks! I've committed your suggestions and will take another pass at the docstring to make similar changes in places that could benefit from it.

> Great @scott-huberty ! I'll give it a try no my EyeLink data as well and report back! 🙂

Sounds good! 🙂

scott-huberty replied:
changed `filename` to `fname` in commit [87320fb](https://github.com/mne-tools/mne-python/pull/11719/commits/87320fb0e2e7a3df63a0efef2193df171a75970c)

mscheltienne commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+    ----------
+    filename : str
+        Path to the eyelink file (.asc).
+    screen_size : tuple
```
```suggestion
    screen_size : tuple of shape (2,)
```

For sequence, it's nice to specify the expected shape.

mscheltienne commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+        Path to the eyelink file (.asc).
+    screen_size : tuple
+        The width and height (in meters) of the screen that the eyetracking
+        data was collected with. For example (.531, .298) for a monitor with
```
```suggestion
        data was collected with. For example ``(.531, .298)`` for a monitor with
```

Double backticks around variables improve readability of the documentation.

mscheltienne commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+        Defaults to None.
+    screen_resolution : tuple
+        The resolution (in pixels) of the screen that the eyetracking data
+        was collected with. For example, (1920, 1080) for a 1920x1080
```
```suggestion
        was collected with. For example, ``(1920, 1080)`` for a 1920x1080
```

mscheltienne reviewed (commented):
A couple of minor comments after gazing through the documentation, which can be applied to other docstrings too. Looks great!

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+
+    Parameters
+    ----------
+    fname : str
```
Should accept a `path-like`

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+    ----------
+    fname : str
+        Path to the eyelink file (.asc).
+    screen_size : tuple of shape (2,)
```
Could we accept an `array-like` and cast to `tuple` if needed?

scott-huberty replied:
Sure, that's a good idea.

This information is just stored in the `Calibration` class and displayed for the user via the `__repr__`  method -  but right now there aren't any functions in `mne` that actually do anything with `screen_size` / `screen_resolution`.  So at the moment I'm not sure if it's even necessary to cast them as `tuple`'s?

We plan to write a function named something like `px_to_degrees`, that would use these values, so maybe we should decide then on whether to automatically caste `screen_size`/`screen_resolution` as `tuple`'s?

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+    screen_distance : float
+        The distance (in meters) from the participant's eyes to the screen.
+        Defaults to None.
+    screen_resolution : tuple of shape (2,)
```
Like above, can we accept `array-like` here?

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+        A list of :class:`mne.preprocessing.eyetracking.Calibration` instances, one for
+        each eye of every calibration that was performed during the recording session.
+    """
+    fname = Path(fname)
```
I believe we actually have a convenience function for this, which also checks if the file exists. Must be somewhere in `utils`... `check_fname` or so? I don't remember

scott-huberty replied:
I saw that! I was going to ask if we could use that. I can try using that instead and push a commit.

larsoner replied:
Probably better to use `name='fname'` because that's the name of the argument in your function

scott-huberty replied:
Thanks both - I incorporated this and @hoechenberger 's other suggestions in [5015c9b](https://github.com/mne-tools/mne-python/pull/11719/commits/5015c9b4e581fef4428d01cc5e203f0cdf131904)

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+    logger.info("Reading calibration data from {}".format(fname))
+    with fname.open() as file:
+        lines = file.readlines()
+        return _parse_calibration(
+            lines, screen_size, screen_distance, screen_resolution
+        )
```
Should be dedented

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -284,6 +287,45 @@ def _find_overlaps(df, max_time=0.05):
+    if not fname.exists():
+        raise FileNotFoundError(f"Could not find file {fname}")
+    logger.info("Reading calibration data from {}".format(fname))
+    with fname.open() as file:
+        lines = file.readlines()
```
I **think** you could do this via `fname.read_text()`.

In either case, please pass the `encoding`

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -328,6 +374,24 @@ def read_raw_eyelink(
+        If True, returns a tuple of (raw, calibrations) where calibrations is
+        a list of Calibration instances, each containing information about a
+        single calibration collected during the recording.
+    screen_size : tuple
```
Same as above,  an this be `array-like`?

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -328,6 +374,24 @@ def read_raw_eyelink(
+        Only set if 'return_calibration' is set to True.
+        The distance from the participant's eyes to the screen in meters.
+        Defaults to None.
+    screen_resolution : tuple
```
Same as above,  an this be `array-like`?

hoechenberger commented on mne/io/eyelink/eyelink.py:
```diff
@@ -357,6 +421,13 @@ def read_raw_eyelink(
     )
+    if return_calibration:
+        calibrations = read_eyelink_calibration(
+            fname, screen_size, screen_distance, screen_resolution
```
Could you please pass more parameters by keyword here?

larsoner commented:
Ignore the `pre` CI failure there's a fix in https://github.com/mne-tools/mne-python/pull/11724

scott-huberty commented on doc/reading_raw_data.rst:
I just want to highlight that I added `read_eyelink_calibration` under the `autosummary` directive in `reading_raw_data.rst`: It seems like I am the first one to add a function that doesn't follow the  `read_raw_*` pattern in this file, so I want to make sure that it is okay for me to do this.

larsoner replied:
mne.io is really just for reading raw data (like samples and such). We might want this in mne.preprocessing.eyetracking or so instead. But we can also change it later as long as it's before we release 1.5 in a couple of months

scott-huberty replied:
No problem, that makes sense - since it is fresh on my mind now I can try to move it to `preprocessing.eyetracking`  before this PR is merged (I agree that it makes sense to keep all the eye-tracking related functions there)

larsoner replied:
Go for it!

scott-huberty replied:
Alright! I think I'll move `mne.io.eyelink.read_eyelink_calibration` to `mne.preprocessing.eyetracking.calibration`, so the API will be `from mne.preprocessing.eyetracking import read_eyelink_calibration`.

britta-wstnr commented:
I gave this a try on my own data and it works like a charm! Will the `Calibration` object (with its info on `screen_size` etc) mesh with the raw EyeLink data at some point? Or am I missing that?

scott-huberty commented:
> I gave this a try on my own data and it works like a charm!

Awesome! 🙂

> Will the `Calibration` object (with its info on `screen_size` etc) mesh with the raw EyeLink data at some point? Or am I missing that?

Yeah I think the idea is for it to be used with a function like:

`mne.preprocessing.eyetracking.convert_units(raw, calibrations[0])`

 where `screen_size` / `screen_resolution` / `screen_distance` will be used to convert `pixel-on-screen` data to `visual-degrees` (on screen), or vice-versa.

I think the last piece of information we need for that function is the unit ( and reference?) of the data in `RawEyelink`, so that we know what the unit is in and if we can convert it?

So, in the case that the `RawEyelink` `eyegaze` samples are in pixels:

The `SampleCoordinateSystem` would be `'pixel-on-screen'`, and the `SampleCoordinateUnit` would be `'pixel'` (according to this [draft](https://bids-specification--1128.org.readthedocs.build/en/1128/modality-specific-files/eye-tracking.html) of an eye-tracking BIDS standard). With that information, we could convert the `pixel-on-screen` data to visual degrees.


We talked about adding a new FIFF specification for this information, but I wanted to see how the `Calibration` class shaped up first, to confirm that we want to add a FIFF speicfication for this and store it in the `Raw` object?

larsoner commented:
> We talked about adding a new FIFF specification for this information, but I wanted to see how the Calibration class shaped up first, to confirm that we want to add a FIFF speicfication for this and store it in the Raw object?

Maybe I'm missing something but if we say the data are in `FIFF.FIFF_UNIT_PX` what more do we need? Doesn't this mean they're pixels-on-screen and whatever info is in `Calibration` can get you directly to visual degrees? I'm not clear on the use case that requires more FIFF constants here

scott-huberty commented:
After discussing with @larsoner and @britta-wstnr during office hours today, Eric is right that there is already a`FIFF.FIFF_UNIT_PX` (that I somehow forgot about!), so we don't need to add another FIFF specification. 

per @britta-wstnr 's point, For simplicity, we'll allow the to-be-develoepd conversion function to accept `screen_size`/`screen_distance`/`screen_resolution` directly. This will be done in a future PR.

We also decided that for this PR we should add a demonstration of loading a calibration in the current eye-tracking tutorial, and i'm going to look again at the `Calibration['points']` array to see if it can be split into separate `dict` keys so that it is easier to initialize/extend to future readers for other manufacturers. I'll add these as TODOS to the top of this message.

scott-huberty commented:
@larsoner / @britta-wstnr sorry for the flurry of commits yesterday, I'm still working out some kinks in my local environment to make sure I can build the doc and run the test suite locally, but I mostly have figured it out.

I think I am happy with this PR. Let me know what you think about the design / code and if you need me to make any further changes!


[Here](https://output.circle-artifacts.com/output/job/be367429-c0f5-447c-9b32-b3f6e0d52000/artifacts/0/html/auto_tutorials/preprocessing/90_eyetracking_data.html) is the updated eye-tracking tutorial. I've added a bit about loading the calibration, at the beginning of the tutorial. I think this should give you a good idea of the API and design without having to load a file yourself. 

[Here](https://output.circle-artifacts.com/output/job/be367429-c0f5-447c-9b32-b3f6e0d52000/artifacts/0/html/generated/mne.preprocessing.eyetracking.Calibration.html#mne.preprocessing.eyetracking.Calibration) is the documentation for the `Calibration` class.

[Here](https://output.circle-artifacts.com/output/job/be367429-c0f5-447c-9b32-b3f6e0d52000/artifacts/0/html/generated/mne.preprocessing.eyetracking.read_eyelink_calibration.html#mne.preprocessing.eyetracking.read_eyelink_calibration) is the documentation for `read_eyelink_calibration`

[Here](https://app.codecov.io/gh/mne-tools/mne-python/pull/11719) is the codecov report. If I'm not mistaken It looks like the `mne/preprocessing/eyetrackingcalibration.py` file I added is covered _88%_ and the`mne/io/eyelink/_utils.py` file I added is covered _96%_.

mscheltienne reviewed (commented):
I looked through the tutorial, looks amazing! Just one comment on the annotation used to mark gaps between acquisition blocks:

> Times between blocks will be annotated with bad_rec_gap.

How about using the existing BAD_ACQ_SKIP annot?

scott-huberty commented:
> How about using the existing BAD_ACQ_SKIP annot?

I agree that's a better idea but I'm kicking myself for not thinking of this when we wrote it -_-

This function was written just  before 1.4 I think so changing it would be an API change I think? So I'll let you all advise on that.

larsoner commented:
> This function was written just before 1.4 I think so changing it would be an API change I think? So I'll let you all advise on that.

I consider it a bugfix so I'd say go ahead

Let me know when it's done and I'll look!

scott-huberty commented:
> > This function was written just before 1.4 I think so changing it would be an API change I think? So I'll let you all advise on that.
> 
> 
> 
> I consider it a bugfix so I'd say go ahead
> 
> 
> 
> Let me know when it's done and I'll look!

Done!

Thanks for the tip, @mscheltienne !

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+
+        If the value for a field is not available, use ``np.nan``. See the example below
+        for more details.
+
```
I'm surprised NumpyDoc allows you to get away with this extra newline. But at least by local convention we don't put them between params
```suggestion
```

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+        screen_distance=None,
+        screen_resolution=None,
+    ):
+        super().__init__()
+        self["onset"] = onset
+        self["model"] = model
+        self["eye"] = eye
+        self["avg_error"] = avg_error
+        self["max_error"] = max_error
+        if points is not None:
+            self.set_calibration_array(points)
+        else:
+            self["points"] = points
+        self["screen_size"] = screen_size
+        self["screen_distance"] = screen_distance
+        self["screen_resolution"] = screen_resolution
```
Since we have to call the super init anyway we might as well have it do some work for us
```suggestion
        super().__init__(
            onset=onset,
            model=model,
            eye=eye,
            avg_error=avg_error,
            max_error=max_error,
            screen_size=screen_size,
            screen_distance=screen_distance,
            screen_resolution=screen_resolution,
        )
        if points is not None:
            self.set_calibration_array(points)
        else:
            self["points"] = points
```

scott-huberty replied:
Oh nice, thanks for the tip!

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+            f"  screen resolution: {screen_resolution} pixels\n"
+        )
+
+    def __getattr__(self, name):
+        """Allow dot indexing of dict keys."""
+        if name in self:
+            return self[name]
+        raise AttributeError(
+            f"'{self.__class__.__name__}' object has no attribute '{name}'"
+        )
```
I'd rather not have this at all and explicitly always access by keys. But if you really want to keep this then this seems safer:
```suggestion
    def __getattr__(self, name):
        """Allow dot indexing of dict keys."""
        if name in self:
            return self[name]
        return super().__getattr__(name)
```

scott-huberty replied:
No problem, in that case I won't commit suggestion then and I will remove this custom dunder method on my local.

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+
+        Returns
+        -------
+        info : instance of Calibration
```
```suggestion
        cal : instance of Calibration
```

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+        """
+        return deepcopy(self)
+
+    def set_calibration_array(self, data):
+        """
+        Create a Numpy Array containing data regarding each calibration point.
+
+        This method takes an array-like objects and converts it into a structured numpy
+        array, with field names ``'point_x'``, ``'point_y'``, ``'offset'``,
+        ``'gaze_x'``, and ``'gaze_y'``.
```
I would avoid the structured array. They are often a pain to access and set etc. To me it's cleaner just to have `self['points']` and `self['gaze']` each of shape `(N, 2)` and `self['offset']` of shape `(N,)` for example.

And to me this `set_calibration_data` really seems like an implementation detail of the `read_eyelink_calibration` function. I think we should get rid of the `__init__` of this function altogether and just have `read_eyelink_calibration` set the `Calibration(..., points=..., gaze=..., offset=...)` properly in the first place.

scott-huberty replied:
Yeah.. I liked the way the structured array kept all the info in one place but totally agree that it's a pretty arcane thing. I was struggling to find a similar implementation with `dict` but your suggestion is nice and i'll implement it and remove `set_calibration_array`.

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+                "Calibration points must be a numpy array. Use "
+                "set_calibration_array() to set calibration data."
+            )
+        fig, ax = plt.subplots()
```
```suggestion
        fig, ax = plt.subplots(constrained_layout=True)
```

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+                    va="center",
+                )
+
+        fig.tight_layout()
```
Rather than tight layout I'd prefer to create the figure with `constrained_layout=True`, it almost always works better
```suggestion
```

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,346 @@
+                )
+
+        fig.tight_layout()
+        fig.show() if show else None
```
We have a `plt_show` wrapper for this (that does some other nice stuff)

larsoner reviewed (commented):
Looks like a great start! Got partway through and think I'm at a good stopping spot -- let me know what you think of my ideas and I'll keep looking after any additional changes (tests and tutorial might need to be updated)

scott-huberty commented on mne/io/eyelink/eyelink.py:
```diff
@@ -293,13 +295,17 @@ def read_raw_eyelink(
-    gap_description="bad_rec_gap",
+    gap_description="BAD_ACQ_SKIP",
+    return_calibration=False,
+    screen_size=None,
+    screen_distance=None,
+    screen_resolution=None,
```
Ok yeah I see your conceen. I'll remove `return_calibration` from `read_raw_eyelink`.

scott-huberty commented:
> Looks like a great start! Got partway through and think I'm at a good stopping spot -- let me know what you think of my ideas and I'll keep looking after any additional changes (tests and tutorial might need to be updated)

Thanks @larsoner! I integrated your suggestions in [6ede0af](https://github.com/mne-tools/mne-python/pull/11719/commits/6ede0af45437355c7736e418662a3fda5571533d) and updated the docs / tests to reflect the changes. 

It looks like the CI's are good. Let me know when you've had a chance to continue your review

hoechenberger commented on doc/changes/latest.inc:
```diff
@@ -34,6 +34,7 @@ Bugs
 - Fix bug with PySide6 compatibility (:gh:`11721` by `Eric Larson`_)
 - Fix hanging interpreter with matplotlib figures using ``mne/viz/_mpl_figure.py`` in spyder console and jupyter notebooks`(:gh:`11696` by `Mathieu Scheltienne`_)
 - Fix bug with overlapping text for :meth:`mne.Evoked.plot` (:gh:`11698` by `Alex Rockhill`_)
+- Deprecated ``gap_description`` kwarg of :func:`mne.io.read_raw_eyelink` and changed default value to ``'BAD_ACQ_SKIP'`` (:gh:`11719` by `Scott Huberty`_)
```
This should go under the API changes section
```suggestion
- Deprecated ``gap_description`` keyword argument of :func:`mne.io.read_raw_eyelink` and changed default value to ``'BAD_ACQ_SKIP'`` (:gh:`11719` by `Scott Huberty`_)
```

scott-huberty replied:
Thanks @hoechenberger  I added the deprecation announcement to the API changes section.

scott-huberty commented:
Just FYI The Test windows 3.10 CI is returning an [error](https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=25374&view=logs&j=305851a9-a7bb-55db-0042-7e2b6f48aa1c&t=fb29f20e-3ff8-5a8b-bdc6-f5ac272b39d0&l=69) "no matching distribution found for pandas" during the pip install action.

Not sure if that's a fluke, Let me know if you want me to re-trigger the CI's.

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,289 @@
+            "origin must be 'top-left', 'top-right', 'bottom-left', or 'bottom-right."
+            f" got {origin}"
+        )
+        assert origin in ("top-left", "top-right", "bottom-left", "bottom-right"), msg
```
`assert` should only be used in tests, or to assert something that we know should be true based on how we (internally) have coded something (which I think is the case for the `assert` statements farther up having to do with `ndarray` -- the `__setitem__` should ensure this for us). It should not be used for user input checks. This is because `assert` statements can be skipped by running `python` in an optimized mode. So instead you should use a helper like:
```suggestion
        _check_option('origin', origin, ("top-left", "top-right", "bottom-left", "bottom-right"))
```

scott-huberty replied:
Oh, okay I see. I think I have a few asserts in this code that I should replace then...

is ``_check_ption`` an mne utility function?

larsoner replied:
Yep, make friends with `git grep` and it will tell you :)
```
~/python/mne-python$ git grep _check_option
...
mne/baseline.py:from .utils import logger, verbose, _check_option, _validate_type
mne/baseline.py:        _check_option(
mne/beamformer/_compute_beamformer.py:    _check_option,
mne/beamformer/_compute_beamformer.py:    _check_option("pick_ori", pick_ori, ("normal", "max-power", "vector", None))
```
```

scott-huberty replied:
hot tip!

larsoner commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,289 @@
+            value = np.array(value)
+        super().__setitem__(key, value)
+
+    def plot(self, title=None, show_offsets=False, origin="top-left", show=True):
```
In the tutorial you do this and in general it's more informative so my vote would be for `show_offsets=True` by default
```suggestion
    def plot(self, title=None, show_offsets=True, origin="top-left", show=True):
```

scott-huberty replied:
good idea!

larsoner reviewed (commented):
Almost there! Just some minor cleanups here and there then we should be good

scott-huberty commented on mne/io/eyelink/eyelink.py:
```diff
@@ -318,15 +320,19 @@ def read_raw_eyelink(
+        times of the left and right eyes are separated by less than 50 ms, then the
+        blink will be merged into a single :class:`mne.Annotations`.
+    gap_description : str (default 'BAD_ACQ_SKIP')
         If there are multiple recording blocks in the file, the description of
```
Okay I'll commit this and then delete the rest of the docstring.

scott-huberty commented on mne/io/eyelink/eyelink.py:
```diff
@@ -418,7 +430,8 @@ class RawEyelink(BaseRaw):
         Whether whether a single eye was tracked ('monocular'), or both
         ('binocular').
     _gap_desc : str
```
Okay, I'll go through and remove these from the `read_raw_eyelink` docstring.

scott-huberty commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,289 @@
+        display.
+    """
+
+    def __init__(
+        self,
+        onset=None,
+        model=None,
+        eye=None,
+        avg_error=None,
+        max_error=None,
+        positions=None,
+        offsets=None,
+        gaze=None,
+        screen_size=None,
+        screen_distance=None,
+        screen_resolution=None,
+    ):
+        super().__init__(
+            onset=onset,
+            model=model,
+            eye=eye,
+            avg_error=avg_error,
+            max_error=max_error,
+            screen_size=screen_size,
+            screen_distance=screen_distance,
+            screen_resolution=screen_resolution,
+        )
+        self["positions"] = positions
+        self["offsets"] = offsets
+        self["gaze"] = gaze
```
The reason is that the ``_set_item__`` dunder method (that I just added since your last review), checks if the values of these keys are `array-like`, and  if so, converts them to numpy arrays. But if you pass a value to these keys directly in the constructor ``__set_item__`` isn't called (at least when I tested it). 

Am I over complicating things? The idea was that the ``plot`` method expects those keys to be numpy arrays so I was try to make sure they are during the instance construction.

larsoner replied:
> Am I over complicating things? The idea was that the plot method expects those keys to be numpy arrays so I was trying to make sure they are during the instance construction.

For simplicity you could probably assume that no user would/should ever set these, so you can assume they are all `ndarray` of the right shape. You don't currently check for example that they are all shape `(n, 2)` or `(n,)`, so the check is currently incomplete anyway. So maybe one way to go is to put a `.. note::` in the class docstring saying that this class is not meant to be constructed by the user.

Another -- if you want to be more complete -- option is to add a `self._check_consistency` method that checks and sets (if necessary) the values to be `ndarray` of the right shape, checking to make sure the shapes are okay. That way you can avoid overriding the `__setitem__` and just call `self._check_consistency()` at the end of `__init__` and also maybe at the start of `self.plot`.

scott-huberty replied:
> So maybe one way to go is to put a `.. note::` in the class docstring saying that this class is not meant to be constructed by the user.

Yeah, I already added a `.. note::` telling users that when possible the calibration instance should be created with a helper function.

So in that case and considering your point that my check was not complete anyway, I think I'll remove the custom ``__set_item__`` dunder method, and we can get a lay of the land after this PR is shipped;  if  it ends up that a lot of people need to create `Calibration` instances from scratch and are running into issues then we can create a robust `_check_consistency` routine?

scott-huberty commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,289 @@
+
+    def __repr__(self):
+        """Return a summary of the Calibration object."""
+        onset = self.get("onset", "N/A")
```
Hmm yeah I see. Yeah,  it's easy enough to make the values for most of those keys `None` or `np.nan`, and most of them are just informational and so having values of `np.nan` or `None` is not going to break anything. 

Maybe making the `'positions'`, `'gaze'`, and `'offsets'` keys obligatory could be a little annoying for users who do have the `'model'`  `'avg_error'` and `'max_error'` information but don't have the data for individual points, in which case they would have to do something like `np.empty(n_calibration_points, 2)`.

But I'm not sure yet how often someone will even create one of these instances from scratch instead of using an available helper function like `read_eyelink_calibration`, so it's hard to say if it will be an issue for people?

larsoner replied:
Let's tell them not to create it from scratch and just make sure we do the right thing internally. This is how Forward, SourceSpaces, etc. have worked for the most part for years and it has been okay. So +1 for keeping as simple as possible and adding more checks later if needed

larsoner reviewed (commented):
Just one last idea then we can merge @scott-huberty !

larsoner commented on mne/io/eyelink/_utils.py:
```diff
@@ -0,0 +1,113 @@
+            gaze = np.array([point[3:] for point in points])
+            # create the Calibration instance
+            calibration = Calibration(
+                onset=0 if onset < 0 else onset,  # 0 if calibrated before recording
```
(and assuming `info['meas_date']` is correct, you could actually get a correct/accurate `datetime` from it)

scott-huberty replied:
@larsoner Yeah good idea.. Plus maybe eyelink let's more than 1 calibration exist before recording start and if so the negative offset could help users differentiate between the cals!

scott-huberty commented:
> Just one last idea then we can merge @scott-huberty !

I think I need to update the docstring to scrub any mention of "will be zero if calibration was before recording start". Plus looks like there is some style complaint. push a commit in a moment.

scott-huberty commented on mne/preprocessing/eyetracking/calibration.py:
```diff
@@ -0,0 +1,239 @@
+        show_offsets : bool
+            Whether to display the offset (in visual degrees) of each calibration
+            point or not. Defaults to ``True``.
+        origin : str
```
@larsoner one last question while I am at it anyway.

Speaking of YAGNI - maybe this `origin` param isn't necessary. I was wondering if we should just have an `ax` param to let the user pass an `axes` object in (I believe `plot_sensors` has this param), then if on the off-chance that someone really has a monitor with origin `(0,0)` NOT at the top left, they can just pass an `axes` object in and make the adjustments they need. What do you think?

larsoner replied:
I had the same thought originally when reading the code. Let's assume all origins will be correct/eyelink-ish until proven otherwise, then decide how to handle it gracefully (if at all) then

scott-huberty commented:
@larsoner assuming the tests will run as smoothly here as the did locally, this is ready for your review again. Note that as discussed I added an `axes` kwarg to the `plot` method and removed he `origin` kwarg.

larsoner reviewed (approved):
Awesome, marking for merge-when-green once I eyeball the example (it looked good last I checked)! Thanks in advance @scott-huberty 

*EDIT: All good https://output.circle-artifacts.com/output/job/a37bd176-ceeb-4441-aeb6-44bf33b2d3b1/artifacts/0/html/auto_tutorials/preprocessing/90_eyetracking_data.html*

britta-wstnr commented:
Yay for merging @scott-huberty ! 🎉
